### PR TITLE
extension context binding (#725)

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
Release Notes candidate: ElementLibrary.on_user_init()`

```
import typing as t
from dateutil import parser

import taipy as tp
from taipy.gui import Gui, State
from taipy.gui.extension import ElementLibrary, Element, ElementProperty, PropertyType


class GuiCoreContext:
    _CORE_CHANGED_NAME = "core_changed"
    __PROP_SCENARIO_CONFIG_ID = "config_id"
    __PROP_SCENARIO_DATE = "date"
    __PROP_SCENARIO_LABEL = "label"
    __SCENARIO_PROPS = (__PROP_SCENARIO_CONFIG_ID, __PROP_SCENARIO_DATE, __PROP_SCENARIO_LABEL)
    _ERROR_VAR = "gui_core_error"

    def __init__(self, gui: Gui, core: tp.Core) -> None:
        self.gui = gui
        self.core = core
        self.scenarios: t.Optional[
            t.List[t.Tuple[str, str, int, bool, t.Optional[t.List[t.Tuple[str, str, int, bool, None]]]]]
        ] = None
        self.scenario_configs: t.Optional[t.List[t.Tuple[str, str]]] = None

    def __add_scenarios(self, res, scenarios):
        for scenario in scenarios:
            res.append((scenario.id, scenario.label, 1, scenario.primary, None))
        return res

    def get_scenarios(self):
        if self.scenarios is None:
            self.scenarios = []
            for cycle, scenarios in tp.get_cycles_scenarios():
                if cycle is None:
                    self.__add_scenarios(self.scenarios, scenarios)
                else:
                    self.scenarios.append((cycle.id, cycle.label, 0, False, self.__add_scenarios([], scenarios)))
        return self.scenarios

    def get_scenario_configs(self):
        if self.scenario_configs is None:
            configs = tp.Config.scenarios
            if isinstance(configs, dict):
                self.scenario_configs = [(id, c.name) for id, c in configs.items()]
        return self.scenario_configs

    def create_new_scenario(self, state: State, id: str, action: str, payload: t.Dict[str, str]):
        config_id = payload.get(GuiCoreContext.__PROP_SCENARIO_CONFIG_ID)
        scenario_config = tp.Config.scenarios.get(config_id)
        if scenario_config is None:
            state.assign(GuiCoreContext._ERROR_VAR, f"Invalid configuration id ({config_id})")
            return
        date_str = payload.get(GuiCoreContext.__PROP_SCENARIO_DATE)
        try:
            date = parser.parse(date_str) if isinstance(date_str, str) else None
        except Exception as e:
            state.assign(GuiCoreContext._ERROR_VAR, "Invalid date ({date_str})")
            return
        scenario = tp.create_scenario(scenario_config, date, payload.get(GuiCoreContext.__PROP_SCENARIO_LABEL))
        for k, v in payload.items():
            if k not in GuiCoreContext.__SCENARIO_PROPS:
                scenario._properties[k] = v
        state.assign(GuiCoreContext._ERROR_VAR, "")

    def broadcast_core_changed(self):
        self.gui.broadcast(GuiCoreContext._CORE_CHANGED_NAME, "")


class GuiCore(ElementLibrary):
    __LIB_NAME = "taipy_gui_core"
    __CTX_VAR_NAME = f"__{__LIB_NAME}_Ctx"

    __elts = {
        "scenario_selector": Element(
            "scenario_id",
            {
                "show_add_button": ElementProperty(PropertyType.dynamic_boolean, True),
                "display_cycles": ElementProperty(PropertyType.dynamic_boolean, True),
                "show_primary_flag": ElementProperty(PropertyType.dynamic_boolean, True),
                "scenario_id": ElementProperty(PropertyType.dynamic_string),
                "scenarios": ElementProperty(PropertyType.react, f"{{{__CTX_VAR_NAME}.get_scenarios()}}"),
                "on_scenario_create": ElementProperty(PropertyType.function, f"{{{__CTX_VAR_NAME}.create_new_scenario}}"),
                "configs": ElementProperty(PropertyType.react, f"{{{__CTX_VAR_NAME}.get_scenario_configs()}}"),
                "core_changed": ElementProperty(PropertyType.broadcast, GuiCoreContext._CORE_CHANGED_NAME),
                "error": ElementProperty(PropertyType.react, GuiCoreContext._ERROR_VAR),
            }
        )
    }

    def get_name(self) -> str:
        return GuiCore.__LIB_NAME

    def get_elements(self) -> t.Dict[str, Element]:
        return GuiCore.__elts

    def get_scripts(self) -> t.List[str]:
        return ["lib/taipy-gui-core.js"]

    def on_init(self, gui: Gui) -> t.Optional[t.Tuple[str, t.Any]]:
        return GuiCore.__CTX_VAR_NAME, GuiCoreContext(gui, tp.Core())
    
    def on_user_init(self, state: State):
        state._add_attribute(GuiCoreContext._ERROR_VAR)
        state._gui._bind_var_val(GuiCoreContext._ERROR_VAR, "")

```

### CODE REVIEW

File: src/taipy/gui/gui.py [hunk 1]
```diff
@@ -1557,6 +1563,15 @@ def __init_route(self):
             except Exception as e:  # pragma: no cover
                 if not self._call_on_exception("on_init", e):
                     warnings.warn(f"Exception raised in on_init.\n{e}")
+        for name, libs in self.__extensions.items():
+            for lib in libs:
+                if not isinstance(lib, ElementLibrary):
```
Comment (FabienLelaquais) [inline src/taipy/gui/gui.py:1568]: Can this even happen?

**Review by FabienLelaquais — APPROVED** [review]
👌

File: src/taipy/gui/gui.py (hunk 1 above)
Comment (FredLL-Avaiga) [reply src/taipy/gui/gui.py:1568]: with python anything can happen !

**Review by dinhlongviolin1 — APPROVED** [review]
Well done!

**Pull request merged**